mri/reconstructors/base.py

```diff
@@ -0,0 +1,139 @@
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+# System import
+import warnings
+
+# Package import
+from ..operators.linear.wavelet import WaveletUD2, WaveletN
+from ..optimizers import pogm, condatvu, fista
+from ..optimizers.utils.cost import GenericCost
+
+# Third party import
+from modopt.opt.linear import Identity
+
+
+class ReconstructorBase(object):
+    """ This is the base reconstructor class for reconstruction.
+    This class holds some parameters that are common for all MR Image
+    reconstructors
+    For the Analysis case, finds the solution  for x of:
+        (1/2) * ||F x - y||^2_2 + mu * H (W x)
+
+    For the Synthesis case, finds the solution of:
+        (1/2) * ||F Wt alpha - y||^2_2 + mu * H(alpha)
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F.
+    linear_op: object
+        Defines the linear sparsifying operator W. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+    regularizer_op: operator, (optional default None)
+        Defines the regularization operator for the regularization function H.
+        If None, the  regularization chosen is Identity and the optimization
+        turns to gradient descent.
+    gradient_formulation: str between 'analysis' or 'synthesis',
+        default 'synthesis'
+        defines the formulation of the image model which defines the gradient.
+    grad_class: Gradient class from mri.operators.
+        Points to the gradient class based on the MR Image model and
+        gradient_formulation.
+    init_gradient_op: bool, default True
+        This parameter controls whether the gradient operator must be
+        initialized right now.
+        If set to false, the user needs to call initialize_gradient_op to
+        initialize the gradient at right time before reconstruction
+    verbose: int, optinal default 0
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    """
+    def __init__(self, fourier_op, linear_op, regularizer_op,
+                 gradient_formulation, grad_class, init_gradient_op=True,
+                 verbose=0, **extra_grad_args):
```

Comment:
we need to document `extra_grad_args`, especially since we override the potential `linear_op` in the synthesis case.